File: clients/python/src/pawrly/query.py
"""Streaming query result."""

from collections.abc import Iterator

from .result import QueryResult


class QueryHandle:
    """A streaming query result. Iterate the rows (memory-bounded), or
    :meth:`collect` them into a :class:`QueryResult`. ``id`` is the server query
    id for :meth:`PawrlyClient.cancel` — populated over gRPC, empty over REST.
    """

    def __init__(self, id: str, rows: Iterator[dict], meta: dict) -> None:
        self.id = id
        self._rows = rows
        self._meta = meta

    def __iter__(self) -> Iterator[dict]:
        return iter(self._rows)

    def collect(self) -> QueryResult:
        rows = list(self._rows)
        return QueryResult(
            columns=self._meta["columns"],
            rows=rows,
            row_count=self._meta["row_count"],
            truncated=self._meta["truncated"],
        )

File: clients/typescript/src/query.ts
import type { QueryResult } from "./result.js";

export type Row = Record<string, unknown>;

/** Mutable completion metadata, filled by a transport's row generator as it
 * drains — read once iteration finishes. */
export interface QueryMeta {
  columns: string[];
  rowCount: number;
  truncated: boolean;
}

/**
 * A streaming query result. Async-iterate the rows (memory-bounded over gRPC),
 * or {@link collect} them into a {@link QueryResult}. `id` is the server-assigned
 * query id for {@link PawrlyClient.cancel} — populated over gRPC, empty over REST
 * (which has no query id).
 */
export class QueryHandle implements AsyncIterable<Row> {
  constructor(
    readonly id: string,
    private readonly stream: AsyncIterable<Row>,
    private readonly meta: QueryMeta,
  ) {}

  [Symbol.asyncIterator](): AsyncIterator<Row> {
    return this.stream[Symbol.asyncIterator]();
  }

  async collect(): Promise<QueryResult> {
    const rows: Row[] = [];
    for await (const row of this.stream) {
      rows.push(row);
    }
    return {
      columns: this.meta.columns,
      rows,
      rowCount: this.meta.rowCount,
      truncated: this.meta.truncated,
    };
  }
}
